types/instance.d.ts

```diff
@@ -224,14 +224,14 @@ export interface FastifyInstance<
    * Set the 404 handler
    */
   setNotFoundHandler<RouteGeneric extends RouteGenericInterface = RouteGenericInterface>(
-    handler: (request: FastifyRequest<RouteGeneric, RawServer, RawRequest>, reply: FastifyReply<RawServer, RawRequest, RawReply, RouteGeneric>) => void
+    handler: (request: FastifyRequest<RouteGeneric, RawServer, RawRequest>, reply: FastifyReply<RawServer, RawRequest, RawReply, RouteGeneric>) => void | Promise<any>
   ): void;
 
   /**
    * Set a function that will be called whenever an error happens
    */
   setErrorHandler<TError extends Error = FastifyError, RouteGeneric extends RouteGenericInterface = RouteGenericInterface>(
-    handler: (this: FastifyInstance<RawServer, RawRequest, RawReply, Logger>, error: TError, request: FastifyRequest<RouteGeneric, RawServer, RawRequest>, reply: FastifyReply<RawServer, RawRequest, RawReply, RouteGeneric>) => void
+    handler: (this: FastifyInstance<RawServer, RawRequest, RawReply, Logger>, error: TError, request: FastifyRequest<RouteGeneric, RawServer, RawRequest>, reply: FastifyReply<RawServer, RawRequest, RawReply, RouteGeneric>) => void | Promise<any>
```

Comment:
The `<any>` should be the fastify instance

Reply:
I don't think error handler should return Fastify instance, so may be `<void>` them?

Reply:
Am I missing something or the promise resolved value will be used as the reply payload. If so, then I'll use unknown in place of any, if we don't want to introduce another generic (but it is not needed here since the user will never access the return value of the handler).

Reply:
@fox1t is correct, if the handler returns a promise, then the resolved value of the promise is sent:
https://github.com/fastify/fastify/blob/f4cb706144bbf2ebb67257914f78020126488ba5/lib/reply.js#L520-L523
https://github.com/fastify/fastify/blob/f4cb706144bbf2ebb67257914f78020126488ba5/lib/wrapThenable.js#L19-L23

I based the signature off of `RouteHandler`, which uses `void | Promise<any>`:
https://github.com/fastify/fastify/blob/f4cb706144bbf2ebb67257914f78020126488ba5/types/route.d.ts#L128-L138

Reply:
Then it is fine! :)

Reply:
Upon taking another look at `any` vs. `undefined`, it seems that `reply.send` uses a default of `undefined` through `RouteGeneric['Reply']`:
https://github.com/fastify/fastify/blob/f4cb706144bbf2ebb67257914f78020126488ba5/types/reply.d.ts#L30
https://github.com/fastify/fastify/blob/f4cb706144bbf2ebb67257914f78020126488ba5/types/utils.d.ts#L44
Maybe these handlers, as well as the route handler types (`RouteHandlerMethod` and `RouteHandler`) should also be typed in a similar manner? They all have a `RouteGeneric` type parameter, and the only way I can think of this breaking user code is if someone is calling route handler methods directly. If this sounds good, I can close this PR and make a new one fixing all the handler types to use `RouteGeneric`.

Reply:
Actually it is `unknown` and not undefined. I think we should always use `unknown` in place of `any` since it is a more safe version of the same concept: it represents any object but forces the user to check if a property exists before using it.
Reference [here](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type)

Reply:
I think that after putting unknown here, we can accept this change request.